### PR TITLE
PKI module - more testcases and crl file loading support

### PR DESCRIPTION
**Pull request by perusworld**
added hardware and software keystore file based tests and added support to load crl file in the keystore implementations. the generated keys have 100 year expiry.

### CODE REVIEW

File: pki/src/main/java/org/hyperledger/besu/pki/keystore/AbstractKeyStoreWrapper.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    if (null == crlLocation) {
+      this.crls = null;
+    } else {
+      try (InputStream stream = new FileInputStream(crlLocation.toFile())) {
+        this.crls =
+            CertificateFactory.getInstance(X_509).generateCRLs(stream).stream()
+                .map(X509CRL.class::cast)
+                .collect(Collectors.toList());
+      } catch (final Exception e) {
+        throw new PkiException("Failed to initialize software truststore: " + crlLocation, e);
```
Comment (atoulme) [inline pki/src/main/java/org/hyperledger/besu/pki/keystore/AbstractKeyStoreWrapper.java:44]: Coild this leak sensitive information in logs?

File: pki/src/test/java/org/hyperledger/besu/pki/keystore/CryptoTestUtil.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      LOG.info("NSS library does not seem to be installed!", e);
+    }
+    return false;
+  }
+
+  public static String getNSSLibPath() throws IOException, InterruptedException {
+    String nssLibPath = ""; // default path
+    final String centOS_nssPathCmd =
+        "whereis libnssdbm3 | grep -o \"\\/.*libnssdbm3\\.[0-9a-z]* \" | sed 's/\\/libnssdbm3.*//g'"; // Jenkins
+    // box
```
Comment (atoulme) [inline pki/src/test/java/org/hyperledger/besu/pki/keystore/CryptoTestUtil.java:44]: Jenkins?

File: pki/src/test/resources/keystore/invalidpartner1client1/crl.pem [hunk 3]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN X509 CRL-----
```
Comment (atoulme) [inline pki/src/test/resources/keystore/invalidpartner1client1/crl.pem:1]: I’d rather we generate these during testing.

File: pki/src/main/java/org/hyperledger/besu/pki/keystore/HardwareKeyStoreWrapper.java [hunk 4]
```diff
@@ -159,7 +164,7 @@ public KeyStore getTrustStore() {
     return keystore;
   }
 
-  private Provider getPkcs11Provider(final String config) {
+  public Provider getPkcs11Provider(final String config) {
```
Comment (atoulme) [inline pki/src/main/java/org/hyperledger/besu/pki/keystore/HardwareKeyStoreWrapper.java:167]: Why was this made public?

File: pki/src/test/resources/keystore/invalidpartner1client1/crl.pem (hunk 3 above)
Comment (atoulme) [reply pki/src/test/resources/keystore/invalidpartner1client1/crl.pem:1]: Same for all these files. It woild be easier to maintain going forward.

Comment (lucassaldanha) [reply pki/src/test/resources/keystore/invalidpartner1client1/crl.pem:1]: Idk if generating them dynamically provides that much value. I'm sure these certificates will have valid date long enough that we won't have an issue maintaining them.

What are your maintenance concerns? Just the expiry date or something else?

Comment (atoulme) [reply pki/src/test/resources/keystore/invalidpartner1client1/crl.pem:1]: How were they generated? How are they matched? What are their relationships?
Some of those certificate files are combined certs - a whole chain. From experience, you don’t just conjure those easily.

Comment (perusworld) [reply pki/src/test/resources/keystore/invalidpartner1client1/crl.pem:1]: The scripts to generate them are available here https://github.com/perusworld/besu-isolated-networks/tree/p2p-over-ssl/scripts, for the unit tests this script was used - gen-test-keys.sh. I can add those scripts to this module itself if that helps. We are working on adding support for pkcs11/12/jks keystores to use it in Node Signing, TLS support p2p. So we generating more keys for acceptance tests and dev quick start etc. Due to the complexity I would reccommend to use the scripts. Thanks

Comment (atoulme) [reply pki/src/test/resources/keystore/invalidpartner1client1/crl.pem:1]: Please include instructions on how to regenerate those files at a minimum.

Comment (atoulme) [reply pki/src/test/resources/keystore/invalidpartner1client1/crl.pem:1]: Your repository has no license. Please add a license.

Comment (perusworld) [reply pki/src/test/resources/keystore/invalidpartner1client1/crl.pem:1]: thanks, i have updated accordingly

**Review by atoulme — APPROVED** [review]
Changes look ok to me.

**Pull request merged**